source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -456,13 +464,80 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, object> allocateI
             }
 
             return sb.ToString();
+
+            void UseExplicitParameters(RelatedDocumentTableData data)
+            {
+                var relatedVariablePrefix = $"{data.TableName.ToLower()}_";
+
+                sb.AppendLine();
+                sb.AppendLine("DELETE FROM @references");
+                sb.AppendLine();
+
+                var valueBlocks = data.Related.Select((r, idx) => $"(@{relatedVariablePrefix}{idx}, '{r.relatedTableName}')");
+                sb.Append("INSERT INTO @references VALUES ");
+                sb.AppendLine(string.Join(", ", valueBlocks));
+                sb.AppendLine();
+
+                sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
+                sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT Reference FROM @references)");
+                sb.AppendLine();
+
+                sb.AppendLine($"INSERT INTO [{data.SchemaName}].[{data.TableName}] ([{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}])");
+                sb.AppendLine($"SELECT @{IdVariableName}, '{mapping.TableName}', Reference, ReferenceTable FROM @references t");
+                sb.AppendLine($"WHERE NOT EXISTS (SELECT null FROM [{data.SchemaName}].[{data.TableName}] r WHERE r.[{data.IdColumnName}] = @{IdVariableName} AND r.[{data.RelatedDocumentIdColumnName}] = t.Reference )");
+
+                for (var x = 0; x < data.Related.Length; x++)
+                    parameters.Add(relatedVariablePrefix + x, data.Related[x].relatedDocumentId);
+            }
+
+            void UseTableValuedParameter(RelatedDocumentTableData data)
+            {
+                var tableVariableName = $"{index}__relatedDocumentTableValuedParameter";
+                parameters.AddTable(tableVariableName, CreateRelatedDocumentTableValuedParameter(mapping, data));
+
+                sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
```

Comment:
These 6 lines seem to be the same no matter which approach is taken. Should we extract the sql and parameterize with the table name?

Reply:
I was thinking more about passing a list of column names to a template.  Imagine there is a bug somewhere between line 498 and 504. How obvious it is that that the same bug exists between lines 481 and 487?

Reply:
The line numbers look wrong, but did you mean something like this?

https://github.com/OctopusDeploy/Nevermore/pull/153/commits/acfcf87ffe82c86bd1e7942b699f725c8762fc52#diff-94be40892362f618a2307a0d89bf656db3bcf5ab10f60619ac87954a1800647aR498